File: predict/src/main/java/com/emarsys/predict/PredictResponseMapper.java
package com.emarsys.predict;

import com.emarsys.core.Mapper;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.util.JsonUtils;
import com.emarsys.predict.api.model.Product;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class PredictResponseMapper implements Mapper<ResponseModel, List<Product>> {

    @Override
    public List<Product> map(ResponseModel responseModel) {
        List<Product> result = new ArrayList<>();
        try {
            JSONObject jsonResponse = new JSONObject(responseModel.getBody());
            JSONObject products = jsonResponse.getJSONObject("products");
            JSONObject features = jsonResponse.getJSONObject("features");
            String cohort = jsonResponse.getString("cohort");
            Iterator<String> keys = features.keys();
            while (keys.hasNext()) {
                String logicName = keys.next();
                JSONObject feature = features.getJSONObject(logicName);
                JSONArray productOrder = feature.getJSONArray("items");

                for (int i = 0; i < productOrder.length(); i++) {
                    JSONObject product = products.getJSONObject(productOrder.getJSONObject(i).getString("id"));

                    Map<String, String> productFields = JsonUtils.toFlatMap(product);

                    Product productBuilder = buildProductFromFields(logicName, cohort, productFields);
                    result.add(productBuilder);
                }
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return result;
    }

    private Product buildProductFromFields(String feature, String cohort, Map<String, String> productFields) {
        String msrp = productFields.remove("msrp");
        String price = productFields.remove("price");
        String available = productFields.remove("available");

        Product.Builder productBuilder = new Product.Builder(
                productFields.remove("item"),
                productFields.remove("title"),
                productFields.remove("link"),
                feature,
                cohort);
        productBuilder.categoryPath(productFields.remove("category"));
        if (available != null) {
            productBuilder.available(Boolean.valueOf(available));
        }
        if (msrp != null) {
            productBuilder.msrp(Float.parseFloat(msrp));
        }
        if (price != null) {
            productBuilder.price(Float.parseFloat(price));
        }
        productBuilder.imageUrl(productFields.remove("image"));
        productBuilder.zoomImageUrl(productFields.remove("zoom_image"));
        productBuilder.productDescription(productFields.remove("description"));
        productBuilder.album(productFields.remove("album"));
        productBuilder.actor(productFields.remove("actor"));
        productBuilder.artist(productFields.remove("artist"));
        productBuilder.author(productFields.remove("author"));
        productBuilder.brand(productFields.remove("brand"));
        String year = productFields.remove("year");
        if (year != null) {
            productBuilder.year(Integer.parseInt(year));
        }
        productBuilder.customFields(productFields);
        return productBuilder.build();
    }
}

File: predict/src/main/java/com/emarsys/predict/PredictResponseMapper.kt
package com.emarsys.predict

import com.emarsys.core.Mapper
import com.emarsys.core.Mockable
import com.emarsys.core.response.ResponseModel
import com.emarsys.core.util.JsonUtils.toFlatMapIncludingNulls
import com.emarsys.core.util.log.Logger.Companion.error
import com.emarsys.core.util.log.entry.CrashLog
import com.emarsys.predict.api.model.Product
import org.json.JSONException
import org.json.JSONObject
import java.util.*

@Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
@Mockable
class PredictResponseMapper : Mapper<ResponseModel, List<Product>> {
    override fun map(responseModel: ResponseModel): List<Product> {
        val result: MutableList<Product> = ArrayList()
        try {
            val jsonResponse = JSONObject(responseModel.body)
            if (jsonResponse.has("products")) {
                val products = jsonResponse.getJSONObject("products")
                val features = jsonResponse.getJSONObject("features")
                val cohort = jsonResponse.getString("cohort")

                features.keys().forEach { logicName ->
                    val feature = features.getJSONObject(logicName)
                    val productOrder = feature.getJSONArray("items")
                    for (i in 0 until productOrder.length()) {
                        val productJson = products.getJSONObject(productOrder.getJSONObject(i).getString("id"))
                        val productFields = toFlatMapIncludingNulls(productJson)
                        val product = createProductFromFields(logicName, cohort, productFields)
                        result.add(product)
                    }
                }
            }
        } catch (e: JSONException) {
            error(CrashLog(e))
        }
        return result
    }

    private fun createProductFromFields(feature: String, cohort: String, productFields: Map<String, String?>): Product {
        val mutableProductFields = productFields.toMutableMap()
        val msrp: String? = mutableProductFields.remove("msrp")
        val price: String? = mutableProductFields.remove("price")
        val available: String? = mutableProductFields.remove("available")
        return Product(
                productId = mutableProductFields.remove("item")!!,
                title = mutableProductFields.remove("title")!!,
                linkUrl = mutableProductFields.remove("link")!!,
                feature = feature,
                cohort = cohort,
                categoryPath = mutableProductFields.remove("category"),
                available = available?.toBoolean(),
                msrp = msrp?.toFloat(),
                price = price?.toFloat(),
                imageUrlString = mutableProductFields.remove("image"),
                zoomImageUrlString = mutableProductFields.remove("zoom_image"),
                productDescription = mutableProductFields.remove("description"),
                album = mutableProductFields.remove("album"),
                actor = mutableProductFields.remove("actor"),
                artist = mutableProductFields.remove("artist"),
                author = mutableProductFields.remove("author"),
                brand = mutableProductFields.remove("brand"),
                year = mutableProductFields.remove("year")?.toInt(),
                customFields = mutableProductFields)
    }
}